File: javascript/node-postgres/src/alternatives/pool/example_with_nonconcurrent_connection_pool.js
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import assert from "node:assert";
import { AuroraDSQLPool } from "@aws/aurora-dsql-node-postgres-connector";

const ADMIN = "admin";
const NON_ADMIN_SCHEMA = "myschema";

function createPool(clusterEndpoint, user) {
  return new AuroraDSQLPool({
    host: clusterEndpoint,
    user: user,
    max: 10,
    idleTimeoutMillis: 30000,
    connectionTimeoutMillis: 10000,
  });
}

async function example() {
  const clusterEndpoint = process.env.CLUSTER_ENDPOINT;
  assert(clusterEndpoint, "CLUSTER_ENDPOINT environment variable is not set");
  const user = process.env.CLUSTER_USER;
  assert(user, "CLUSTER_USER environment variable is not set");

  const pool = createPool(clusterEndpoint, user);

  try {
    // Get a client from the pool
    const client = await pool.connect();

    try {
      if (user !== ADMIN) {
        await client.query("SET search_path=" + NON_ADMIN_SCHEMA);
      }

      // Create a new table
      await client.query(`CREATE TABLE IF NOT EXISTS owner (
        id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
        name VARCHAR(30) NOT NULL,
        city VARCHAR(80) NOT NULL,
        telephone VARCHAR(20)
      )`);

      // Insert some data
      await client.query(
        "INSERT INTO owner(name, city, telephone) VALUES($1, $2, $3)",
        ["John Doe", "Anytown", "555-555-1900"]
      );

      // Check that data is inserted by reading it back
      const result = await client.query(
        "SELECT id, city FROM owner where name='John Doe'"
      );
      assert.deepEqual(result.rows[0].city, "Anytown");
      assert.notEqual(result.rows[0].id, null);

      await client.query("DELETE FROM owner where name='John Doe'");
      console.log("Completed successfully");
    } finally {
      // Release the client back to the pool
      client.release();
    }
  } catch (error) {
    console.error(error);
    throw error;
  } finally {
    await pool.end();
  }
}

export { example };

File: python/asyncpg/src/alternatives/pool/example_with_nonconcurrent_connection_pool.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

import asyncio
import os

import aurora_dsql_asyncpg as dsql


async def connect_with_pool(cluster_user, cluster_endpoint):
    ssl_cert_path = "./root.pem"
    if not os.path.isfile(ssl_cert_path):
        raise FileNotFoundError(f"SSL certificate file not found: {ssl_cert_path}")

    pool_params = {
        "user": cluster_user,
        "host": cluster_endpoint,
        "ssl": "verify-full",
        "sslrootcert": ssl_cert_path,
        "min_size": 2,
        "max_size": 5,
    }

    pool = await dsql.create_pool(retry=True, **pool_params)
    try:
        async with pool.acquire() as conn:
            result = await conn.fetchval("SELECT 1")
            assert result == 1

        async def insert_owner(conn):
            await conn.execute(
                "INSERT INTO owner(name, city, telephone) VALUES($1, $2, $3)",
                "John Doe",
                "Anytown",
                "555-555-1900",
            )

        await pool.run_transaction(insert_owner)
    finally:
        await pool.close()


async def main():
    try:
        cluster_user = os.environ.get("CLUSTER_USER", None)
        assert cluster_user is not None, "CLUSTER_USER environment variable is not set"

        cluster_endpoint = os.environ.get("CLUSTER_ENDPOINT", None)
        assert cluster_endpoint is not None, "CLUSTER_ENDPOINT environment variable is not set"

        ssl_cert_path = "./root.pem"
        if not os.path.isfile(ssl_cert_path):
            raise FileNotFoundError(f"SSL certificate file not found: {ssl_cert_path}")

        await connect_with_pool(cluster_user, cluster_endpoint)

    finally:
        pass

    print("Pool exercised successfully")


if __name__ == "__main__":
    asyncio.run(main())
